File: src/com/nadmm/airports/utils/TimeUtils.java
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
 */

package com.nadmm.airports.utils;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.format.DateFormat;
import android.text.format.DateUtils;

import com.nadmm.airports.PreferencesActivity;

public class TimeUtils {

    public static CharSequence formatLongDateTime( long time ) {
        return DateFormat.format( "MMM dd, yyyy h:mmaa", new Date( time ) );
    }

    public static String formatDateTime( Context context, long millis ) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences( context );
        boolean local = prefs.getBoolean( PreferencesActivity.SHOW_LOCAL_TIME, false );
        if ( local ) {
            return formatDateTimeLocal( context, millis );
        } else {
            return formatDateTimeUTC( context, millis );
        }
    }

    public static String formatDateTimeYear( Context context, long millis ) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences( context );
        boolean local = prefs.getBoolean( PreferencesActivity.SHOW_LOCAL_TIME, false );
        if ( local ) {
            return formatDateTimeYearLocal( context, millis );
        } else {
            return formatDateTimeYearUTC( context, millis );
        }
    }

    public static String formatDateRange( Context context, long startMillis, long endMillis ) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences( context );
        boolean local = prefs.getBoolean( PreferencesActivity.SHOW_LOCAL_TIME, false );
        if ( local ) {
            return formatDateRangeLocal( context, startMillis, endMillis );
        } else {
            return formatDateRangeUTC( context, startMillis, endMillis );
        }
    }

    @SuppressWarnings("deprecation")
    public static String formatDateTimeUTC( Context context, long millis ) {
        String s = DateUtils.formatDateRange( context, millis, millis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_NO_YEAR
                | DateUtils.FORMAT_ABBREV_ALL
                | DateUtils.FORMAT_UTC );
        return String.format( "%s UTC", s );
    }

    @SuppressWarnings("deprecation")
    public static String formatDateTimeYearUTC( Context context, long millis ) {
        String s = DateUtils.formatDateRange( context, millis, millis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_ABBREV_ALL
                | DateUtils.FORMAT_UTC );
        return String.format( "%s UTC", s );
    }

    @SuppressWarnings("deprecation")
    public static String formatDateTimeLocal( Context context, long millis ) {
        String s = DateUtils.formatDateRange( context, millis, millis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_NO_YEAR
                | DateUtils.FORMAT_ABBREV_ALL );
        return String.format( "%s %s", s, getLocalTimeZoneName() );
    }

    @SuppressWarnings("deprecation")
    public static String formatDateTimeYearLocal( Context context, long millis ) {
        String s = DateUtils.formatDateRange( context, millis, millis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_ABBREV_ALL );
        return String.format( "%s %s", s, getLocalTimeZoneName() );
    }

    @SuppressWarnings("deprecation")
    public static String formatDateRangeUTC( Context context,
            long startMillis, long endMillis ) {
        String s = DateUtils.formatDateRange( context, startMillis, endMillis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_NO_YEAR
                | DateUtils.FORMAT_ABBREV_ALL
                | DateUtils.FORMAT_UTC );
        return String.format( "%s UTC", s );
    }

    @SuppressWarnings("deprecation")
    public static String formatDateRangeLocal( Context context,
            long startMillis, long endMillis ) {
        String s = DateUtils.formatDateRange( context, startMillis, endMillis,
                DateUtils.FORMAT_24HOUR
                | DateUtils.FORMAT_SHOW_DATE
                | DateUtils.FORMAT_SHOW_TIME
                | DateUtils.FORMAT_NO_YEAR
                | DateUtils.FORMAT_ABBREV_ALL );
        return String.format( "%s %s", s, getLocalTimeZoneName() );
    }

    public static String getLocalTimeZoneName() {
        TimeZone tz = TimeZone.getDefault();
        return tz.getDisplayName( tz.inDaylightTime( new Date() ), TimeZone.SHORT );
    }

    public static CharSequence formatElapsedTime( long time ) {
        Date now = new Date();
        return formatElapsedTime( now.getTime(), time );
    }

    public static CharSequence formatElapsedTime( long time1, long time2 ) {
        return DateUtils.getRelativeTimeSpanString( time2, time1,
                DateUtils.MINUTE_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE );
    }

    public static String getTimeZoneAsString( TimeZone tz ) {
        Date now = new Date();
        String tzName = tz.getDisplayName( tz.inDaylightTime( now ), TimeZone.SHORT );
        SimpleDateFormat tzFormat = new SimpleDateFormat( "'(UTC'Z')'", Locale.US );
        tzFormat.setTimeZone( tz );
        return String.format( "%s %s", tzName, tzFormat.format( now ) );
    }

}

File: src/main/java/com/nadmm/airports/utils/TimeUtils.kt
/*
 * FlightIntel for Pilots
 *
 * Copyright 2011-2025 Nadeem Hasan <nhasan@nadmm.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nadmm.airports.utils

import android.content.Context
import android.text.format.DateUtils
import com.nadmm.airports.ActivityBase
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object TimeUtils {
    private val ISO3339_FORMAT_UTC = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.US)
    private val ISO3339_FORMAT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZZZZZ", Locale.US)
    private val ISO3339_MILLIS_FORMAT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ", Locale.US)
    private val FAA_FORMAT = SimpleDateFormat("MM/dd/yyyy", Locale.US)

    init {
        ISO3339_FORMAT_UTC.timeZone = TimeZone.getTimeZone("UTC")
        FAA_FORMAT.timeZone = TimeZone.getTimeZone("UTC")
    }

    fun formatDateTime(activity: ActivityBase, millis: Long): String {
        if (activity.prefShowLocalTime) {
            return formatDateTimeLocal(activity, millis)
        } else {
            return formatDateTimeUTC(activity, millis)
        }
    }

    fun formatDateTimeYear(activity: ActivityBase, millis: Long): String {
        if (activity.prefShowLocalTime) {
            return formatDateTimeYearLocal(activity, millis)
        } else {
            return formatDateTimeYearUTC(activity, millis)
        }
    }

    fun formatDateRange(activity: ActivityBase, start: Calendar, end: Calendar): String {
        return formatDateRange(activity, start.getTimeInMillis(), end.getTimeInMillis())
    }

    fun formatDateRange(activity: ActivityBase, startMillis: Long, endMillis: Long): String {
        if (activity.prefShowLocalTime) {
            return formatDateRangeLocal(activity, startMillis, endMillis)
        } else {
            return formatDateRangeUTC(activity, startMillis, endMillis)
        }
    }

    @Suppress("deprecation")
    fun formatDateTimeUTC(context: Context?, millis: Long): String {
        val s = DateUtils.formatDateRange(
            context, millis, millis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_NO_YEAR
                    or DateUtils.FORMAT_ABBREV_ALL
                    or DateUtils.FORMAT_UTC)
        )
        return "$s UTC"
    }

    @Suppress("deprecation")
    fun formatDateTimeYearUTC(context: Context?, millis: Long): String {
        val s = DateUtils.formatDateRange(
            context, millis, millis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_ABBREV_ALL
                    or DateUtils.FORMAT_UTC)
        )
        return String.format("%s UTC", s)
    }

    @Suppress("deprecation")
    fun formatDateTimeLocal(context: Context?, millis: Long): String {
        val s = DateUtils.formatDateRange(
            context, millis, millis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_NO_YEAR
                    or DateUtils.FORMAT_ABBREV_ALL)
        )
        return String.format("%s %s", s, localTimeZoneName)
    }

    @Suppress("deprecation")
    fun formatDateTimeYearLocal(context: Context?, millis: Long): String {
        val s = DateUtils.formatDateRange(
            context, millis, millis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_ABBREV_ALL)
        )
        return String.format("%s %s", s, localTimeZoneName)
    }

    @Suppress("deprecation")
    fun formatDateRangeUTC(
        context: Context?,
        startMillis: Long, endMillis: Long
    ): String {
        val s = DateUtils.formatDateRange(
            context, startMillis, endMillis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_NO_YEAR
                    or DateUtils.FORMAT_ABBREV_ALL
                    or DateUtils.FORMAT_UTC)
        )
        return String.format("%s UTC", s)
    }

    @Suppress("deprecation")
    fun formatDateRangeLocal(
        context: Context?,
        startMillis: Long, endMillis: Long
    ): String {
        val s = DateUtils.formatDateRange(
            context, startMillis, endMillis,
            (DateUtils.FORMAT_24HOUR
                    or DateUtils.FORMAT_SHOW_DATE
                    or DateUtils.FORMAT_SHOW_TIME
                    or DateUtils.FORMAT_NO_YEAR
                    or DateUtils.FORMAT_ABBREV_ALL)
        )
        return String.format("%s %s", s, localTimeZoneName)
    }

    val localTimeZoneName: String
        get() {
            val tz = TimeZone.getDefault()
            return tz.getDisplayName(tz.inDaylightTime(Date()), TimeZone.SHORT)
        }

    fun formatElapsedTime(time: Long): CharSequence? {
        val now = Date()
        return formatElapsedTime(now.getTime(), time)
    }

    fun formatElapsedTime(time1: Long, time2: Long): CharSequence? {
        if ((time1 - time2) < DateUtils.MINUTE_IN_MILLIS) {
            return "just now"
        }
        return DateUtils.getRelativeTimeSpanString(
            time2, time1,
            DateUtils.MINUTE_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE
        )
    }

    fun getTimeZoneAsString(tz: TimeZone): String {
        val now = Date()
        val tzName = tz.getDisplayName(tz.inDaylightTime(now), TimeZone.SHORT)
        val tzFormat = SimpleDateFormat("'(UTC'Z')'", Locale.US)
        tzFormat.timeZone = tz
        return String.format("%s %s", tzName, tzFormat.format(now))
    }

    @JvmStatic
    fun format3339(date: Date): String {
        return ISO3339_FORMAT_UTC.format(date) + "Z"
    }

    @JvmStatic
    fun parse3339(s: String): Date? {
        // This is needed as SimpleDateFormat does not parse RFC3339 "Z" for UTC.
        // Convert to ISO8601 format first if using "Z"
        val iso8601 = s.replace("Z$".toRegex(), "+00:00")
        var date: Date?
        try {
            date = ISO3339_FORMAT.parse(iso8601)
        } catch (_: ParseException) {
            date = null
        }

        if (date == null) {
            try {
                date = ISO3339_MILLIS_FORMAT.parse(iso8601)
            } catch (_: ParseException) {
            }
        }

        return date
    }

    fun parseFaaDate(s: String): Date? {
        var date: Date?
        try {
            date = FAA_FORMAT.parse(s)
        } catch (_: ParseException) {
            date = null
        }

        return date
    }
}
